### PR TITLE
add mask_convert_to_lp to support bool->fp16/bf16 convert

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Fix
https://github.com/pytorch/pytorch/issues/117624
https://github.com/pytorch/pytorch/issues/117627

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117830

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/codegen/cpp_prefix.h [hunk 1]
```diff
@@ -364,6 +364,12 @@ inline at::vec::Vectorized<float> mask_convert_to_float(at::vec::Vectorized<floa
   return at::vec::Vectorized<float>::blendv(zeros, ones, src);
 }
 
+template <typename scalar_t>
+inline at::vec::Vectorized<scalar_t> mask_convert_to_lp(at::vec::Vectorized<float> src) {
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_prefix.h:368]:
```suggestion
template <typename scalar_t>
inline
typename std::enable_if<std::is_same<scalar_t, bfloat16>::value || std::is_same<scalar_t, half>::value, at::vec::Vectorized<scalar_t>>::type
mask_convert_to_lowp(at::vec::Vectorized<float> src) {
```

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -1345,6 +1345,8 @@ def to_dtype(x, dtype, src_dtype=None):
             return f"vec_convert_to_mask({x})"
         if opt_ctx_x.dtype == torch.bool and dtype in (torch.float, torch.float32):
             return f"mask_convert_to_float({x})"
+        if opt_ctx_x.dtype == torch.bool and dtype in DTYPE_LOWP_FP:
+            return f"mask_convert_to_lp<{DTYPE_TO_CPP[dtype]}>({x})"
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1349]:
```suggestion
            return f"mask_convert_to_lowp<{DTYPE_TO_CPP[dtype]}>({x})"
```

File: test/inductor/test_cpu_repro.py [hunk 3]
```diff
@@ -2765,6 +2765,22 @@ def forward(self, x):
                 "Vectorized<float>::loadu(tmpbuf.data())", 0, exactly=True
             ).run(code)
 
+    def test_logical_op_store_to_lp_data_dtype(self):
```
Comment (jgong5) [inline test/inductor/test_cpu_repro.py:2768]:
```suggestion
    def test_logical_op_store_to_lowp_data_dtype(self):
```

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: test/inductor/test_cpu_repro.py (hunk 3 above)
Comment (zhuhaozhe) [reply test/inductor/test_cpu_repro.py]: thanks for advice, all revised.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**